substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/AnalysisMethod.java

```diff
@@ -258,6 +259,28 @@ public void registerAsInlined() {
         return getInvokeLocations().stream().map(location -> (AnalysisMethod) location.getMethod()).collect(Collectors.toSet());
     }
 
+    public static class CallContext {
```

Comment:
Why is this class necessary?

Reply:
Currently,  `CallTreePrinter` uses `MethodTypeFlow` and `InvokeTypeFlow`, which are points-to specific classes. The CallContext was my first attempt to create an abstraction around it. However, I was not particularly happy about it, so I will be grateful for your feedback & suggestions. I can think of the following ways.

1. Create a helper method on `AnalysisMethod` providing the information necessary for building the call tree. This is essentially what I was trying to achieve with `CallContext` and `getCallees` - get the same kind of information, but without using points-to specific classes.
2. Move the whole call-tree building algorithm behind an abstraction, so that different types of analysis can construct it in different ways. We could for example have an abstraction to create the full call graph and then another algorithm to convert the call graph into a call tree. The call graph to call tree conversion could already be shared.
3. Leave it be for now and let the `CallTreePrinter` be points-to dependent. This would of course be the easiest option, but I will (hopefully soon) have to solve this problem somehow anyway, so I prefer to resolve it now then later.

What is your opinion? I believe the second approach could be the cleanest. It will require more work now, but having an API for creating call graphs might be useful for other purposes as well (finding paths to 'unwanted methods' or between different methods, some lightweight checks, code smell detection, etc) and I believe there are other places across the project that generate call graphs, the [PermissionsFeature](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.truffle.tck/src/com/oracle/svm/truffle/tck/PermissionsFeature.java) for example, and I could try to design the API in such a way that these clients can use it instead of having their own call graph creation algorithms.

Reply:
After discussions with @vjovanov, we decided to follow option 3 and just leave it be for now if no one objects. When the time comes to implement the call tree printing for Reachability Analysis, we will know more about what the shared API could look like and therefore we should be able to make a better decision.